pkg/resources/provisioning.cattle.io/v1/cluster/validator.go

```diff
@@ -771,6 +783,129 @@ func (p *provisioningAdmitter) validateS3Secret(oldCluster, cluster *v1.Cluster)
+// (a) references an existing ETCDSnapshot, and
+// (b) contains decodable metadata with a valid "provisioning-cluster-spec".
+func (p *provisioningAdmitter) validateETCDSnapshotRestore(request *admission.Request, oldCluster, newCluster *v1.Cluster) (*admissionv1.AdmissionResponse, error) {
+	if request.Operation != admissionv1.Update {
+		return admission.ResponseAllowed(), nil
+	}
+
+	// No RKEConfig means no restore spec.
+	if newCluster.Spec.RKEConfig == nil {
+		return admission.ResponseAllowed(), nil
+	}
+
+	newRestore := newCluster.Spec.RKEConfig.ETCDSnapshotRestore
+
+	// Allow if restore spec is being cleared or is empty.
+	if newRestore == nil || newRestore.Name == "" || newRestore.RestoreRKEConfig == "" {
+		return admission.ResponseAllowed(), nil
+	}
+
+	var oldRestore *rkev1.ETCDSnapshotRestore
+	if oldCluster.Spec.RKEConfig != nil {
+		oldRestore = oldCluster.Spec.RKEConfig.ETCDSnapshotRestore
+	}
+
+	// Allow if spec is unchanged, to avoid blocking unrelated cluster updates.
+	if equality.Semantic.DeepEqual(oldRestore, newRestore) {
+		return admission.ResponseAllowed(), nil
+	}
+
+	snap, err := p.etcdSnapshotCache.Get(newCluster.Namespace, newRestore.Name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return admission.ResponseBadRequest(
+				fmt.Sprintf("etcd restore references missing snapshot %s in namespace %s", newRestore.Name, newCluster.Namespace)), nil
+		}
+		return nil, fmt.Errorf("failed to get etcd snapshot %s/%s: %w", newCluster.Namespace, newRestore.Name, err)
+	}
+
+	var clusterSpec *v1.ClusterSpec
+	var decodeErr error
+	// Only parse snapshot metadata if the restore mode requires it.
+	if newRestore.RestoreRKEConfig != "none" {
+		clusterSpec, decodeErr = parseSnapshotClusterSpec(snap)
+		if decodeErr != nil {
+			return admission.ResponseBadRequest(
+				fmt.Sprintf("invalid ETCD snapshot metadata for %s/%s: %v", snap.Namespace, snap.Name, decodeErr)), nil
+		}
+	}
+
+	switch newRestore.RestoreRKEConfig {
+	case "none":
+		return admission.ResponseAllowed(), nil
+
+	case "kubernetesVersion":
+		if clusterSpec.KubernetesVersion == "" {
+			return admission.ResponseBadRequest("snapshot metadata missing KubernetesVersion for kubernetesVersion restore"), nil
+		}
+		return admission.ResponseAllowed(), nil
+
+	case "all":
+		if clusterSpec.RKEConfig == nil || clusterSpec.KubernetesVersion == "" {
+			return admission.ResponseBadRequest("snapshot metadata must include RKEConfig and KubernetesVersion for 'all' restore"), nil
+		}
+		return admission.ResponseAllowed(), nil
+
+	default:
+		return admission.ResponseBadRequest(fmt.Sprintf("unsupported restore mode %s", newRestore.RestoreRKEConfig)), nil
+	}
+}
+
+// parseSnapshotClusterSpec decodes snapshot.SnapshotFile.Metadata into a v1.ClusterSpec.
+// The metadata is stored as a nested, gzipped, base64-encoded structure.
+func parseSnapshotClusterSpec(snap *rkev1.ETCDSnapshot) (*v1.ClusterSpec, error) {
```

Comment:
We copy this from Rancher correct? I wonder if we could move this to `github.com/rancher/rancher/pkg/apis/provisioning.cattle.io/v1` somewhere.